slides/k8s/operators.md

```diff
@@ -0,0 +1,366 @@
+# Operators
+
+- Operators are one of the many ways to extend Kubernetes
+
+- We will define operators
+
+- We will see how they work
+
+- We will install a specific operator (for ElasticSearch)
+
+- We will use it to provision an ElasticSearch cluster
+
+---
+
+## What are operators?
+
+- Operators combine two things:
+
+  - Custom Resource Definitions
+
+  - controller code watching the corresponding resources and acting upon them
+
+- A given operator can define one or multiple CRDs
+
+- The controller code (control loop) can run within the cluster or elsewhere
+
+  (running as a Deployment with 1 replica is a common scenario)
+
+---
+
+## Why use operators?
+
+- Kubernetes gives us Deployments, StatefulSets, Services ...
+
+- These mechanisms give us building blocks to deploy applications
+
+- They work great for services that are made of *N* identical containers
+
+  (like stateless ones)
+
+- They also work great for some stateful applications like Consul, etcd ...
+
+  (with the help of highly persistent volumes)
+
+- They're not enough for complex services:
+
+  - where different containers have different roles
+
+  - where extra steps have to be taken when scaling or replacing containers
+
+---
+
+## Use-cases for operators
+
+- Systems with primary/secondary replication
+
+  Examples: MariaDB, MySQL, PostgreSQL, Redis ...
+
+- Systems where different groups of nodes have different roles
+
+  Examples: ElasticSearch, MongoDB ...
+
+- Systems that depend on other non-trivial systems
```

Comment:
"other non-trivial systems" could be _anything_ - is it reasonable to scope it more specifically?

Reply:
Tricky! It could really be anything. The only examples I found so far were the ones I listed (Flink and Kafka), who both depend on Zookeeper.

But we could also imagine a PostgreSQL operator that would depend on an Etcd cluster for leader election; a Graylog operator that would depend on Mongo, Redis, and other operators, etc.

Reply:
Hmm, maybe "other systems which could have specific dependencies"?

Reply:
OK, here is a new wording; I hope it is better.

- Systems with complex dependencies (that are themselves managed with operators)

  Examples: Flink or Kafka, which both depend on Zookeeper

Reply:
That's much better/more clear.